blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc

```diff
@@ -0,0 +1,647 @@
+
+== Solution Overview
+
+The following image shows the overall solution design, based on the example of a microservice for managing vegetable data:
+
+++++
+<div class="imageblock centered-image">
+    <img src="/images/auditing_overview.png" style="max-width:100%; margin-bottom:10px; margin-top:10px;" class="responsive-image" alt="Auditing With Change Data Capture and Stream Processing">
+</div>
+++++
+
+The architecture is made up of these components:
+
+* _vegetables-service_: a simple REST service for inserting and updating vegetable data into a Postgres database;
+as part of its processing, it will not only update its actual "business table" `vegetable`,
+but also insert some auditing metadata into a dedicated metadata table `transaction_context_data`;
+Debezium is used to stream change events from the two tables into corresponding topics in Apache Kafka
+* _log-enricher_: a stream processing based on Kafka Streams,
+which enriches the messages from the CDC topic containing the `vegetable` change events (`dbserver1.inventory.vegetable`) with the corresponding metadata in the `dbserver1.inventory.transaction_context_data` topic sourced from the `transaction_context_data` table and writes the enriched vegetable change event back to Kafka into the `dbserver1.inventory.vegetable.enriched` topic.
+
+You can find a https://github.com/debezium/debezium-examples/tree/master/auditlog[complete example] with all the components on GitHub.
+
+== Providing Auditing Metadata
+
+Let's first discuss how an application such as the vegetable service can provide the required auditing metadata.
+As an example, the following metadata should be made available for auditing purposes:
+
+* The application user that did a data change, as represented by the `sub` claim of a JWT token (https://tools.ietf.org/html/rfc7519[JSON Web Token])
+* The request timestamp, as represented by the `Date` HTTP header
+* A use case identifier, as provided via a custom Java annotation on the invoked REST resource method
+
+Here is a basic implementation of a REST resource for persisting a new vegetable using the https://jcp.org/en/jsr/detail?id=370[JAX-RS API]:
+
+[source,java]
+----
+@Path("/vegetables")
+@RequestScoped
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+public class VegetableResource {
+
+    @Inject
+    VegetableService vegetableService;
+
+    @POST
+    @RolesAllowed({"farmers"})
+    @Transactional
+    @Audited(useCase="CREATE VEGETABLE")
+    public Response createVegetable(Vegetable vegetable) {
+        if (vegetable.getId() != null) {
+            return Response.status(Status.BAD_REQUEST.getStatusCode()).build();
+        }
+
+        vegetable = vegetableService.createVegetable(vegetable);
+
+        return Response.ok(vegetable).status(Status.CREATED).build();
+    }
+
+    // update, delete ...
+}
+----
+
+Not too exciting, if you've ever built REST services with JAX-RS before:
```

Comment:
I'd rephrase that somehow even though you are right about the "excitment" this might give readers :-)

Reply:
Yes, good point. It didn't feel quite right to me, too, but I couldn't quite say what it is. Will reword.